### ISSUE TITLE
"Incremental" LogiX Node

### ISSUE THREAD

**Issue by 3x1t-5tyl3** [Neos-Metaverse/NeosPublic#2152]
### ## Is your feature request related to a problem? Please describe.
Currently we do speed for things (like visualisers) through using T and multiplying it's output. This creates an issue where:
if the session is open for 2000 seconds and you multiply that by 0,5 you get 1000, thus halfing the value instead of halfing the speed creating several issues overtime including floatingpoint errors aswell as creating issues when tryign to adjust speed making things switch quickly.

## Relevant issues
-

## Describe the solution you'd like
An Incremental node that's non-relient on time and can be increased it's speed without changing the value inherently.
Input:
Impulse (to reset to 0)
Speed multiplier (float)

Output:
An incremental output (float) changing it's value by +1 if multiply is set to 1 per second
+.5 if multiply is set to 0.5, +2 if set to 2.

Example: If I plug in 0.1 as a multiplier the value changes by 0.1 per second. So in 10 seconds I have a value of 1; After those 10 seconds I decide to switch the multiplier to "1" and another 10 seconds later the value output on the node should be "11" (1 from previous + 10 from the new multiplier)

## Describe alternatives you've considered
Stopwatch and drive the speed on that instead. Which is not a great idea. Otherwise valuegradient drivers and a bunch of other non-sense which still doesn't quite eleveate the issue regarding having to *multiply* a number instead of changing the speed of it's incrementation.

This is terrifingly bad and has floating point errors and requires alot of impulses.
![grafik](https://user-images.githubusercontent.com/61988002/116952074-0a1bf200-ac8a-11eb-93f6-ffe393ab8073.png)

## Additional context
https://user-images.githubusercontent.com/61988002/116951735-ef954900-ac88-11eb-90d7-0813f50d0378.mp4

**Comment by shadowpanther** [comment]
This is how I usually do it in a totally local way as a workaround:
![image](https://user-images.githubusercontent.com/675104/116954114-bfa17200-ac97-11eb-8574-9a5812b56c60.png)

**Comment by Earthmark** [comment]
I did end up creating similar kinds of programs during the original movement system for iron scavengers, where a node had to have 'checkpoints' where it would update the rate of change relative to a time.

The pattern I ended up using was to store the last time the value was updated and the value at that time, then using an OnChange node to update the two when rate changed.

I bring this up because I was worried about the update cost of the first proposed resolution, where I am not confident that circuit would be reliable under network latency conditions (it might also have a race condition where user machine all fight to update the register? but I'm not sure of that). 

The local approach seems like a possible alternative, does this particular circuit need to be network sync'd for your purposes Exit?

The behavior I had with my circuit was almost like a lerp with no target, only an accumulator value. However, to my knowledge the lerps don't sync their internal value, I believe it's assumed that values will converge anyways, so there's no need. That this is a node that introduces a network sync, that might get more interesting. If a network sync isn't done, I could see the accumulator getting desynchronized between different users over time, as tiny errors in when the rate changed grow.
Or the value wouldn't get populated for a newly joining user, resulting in them being dramatically off without a network sync.

**Comment by 3x1t-5tyl3** [comment]
> 
> 
> I did end up creating similar kinds of programs during the original movement system for iron scavengers, where a node had to have 'checkpoints' where it would update the rate of change relative to a time.
> 
> The pattern I ended up using was to store the last time the value was updated and the value at that time, then using an OnChange node to update the two when rate changed.
> 
> I bring this up because I was worried about the update cost of the first proposed resolution, where I am not confident that circuit would be reliable under network latency conditions (it might also have a race condition where user machine all fight to update the register? but I'm not sure of that).
> 
> The local approach seems like a possible alternative, does this particular circuit need to be network sync'd for your purposes Exit?
> 
> The behavior I had with my circuit was almost like a lerp with no target, only an accumulator value. However, to my knowledge the lerps don't sync their internal value, I believe it's assumed that values will converge anyways, so there's no need. That this is a node that introduces a network sync, that might get more interesting. If a network sync isn't done, I could see the accumulator getting desynchronized between different users over time, as tiny errors in when the rate changed grow.
> Or the value wouldn't get populated for a newly joining user, resulting in them being dramatically off without a network sync.
Yeah the value should be synced. The whole idea here is to have a way to have a number than increases without the adverse effect of multiplying T where T would become unwieldly quickly. So yes network synced is wanted. Should've noted that down.

**Comment by Frooxius** [comment]
There's a few ways this could be approached. How often are you varying the speed? If you only change it occasionally (e.g. on Impulse) a modified version of Stopwatch node for example with Speed input would work well, with minimal added network traffic.

Alternatively we can add node that does sync the accumulated value, but this might be a bit trickier, as it'll depend on the host handling that (or another user) and will depend on their update rate without any sort of smoothing or adjusting, though that's relatively easy to fix. There are mechanism in Neos that do this kind of thing, like AuthorityTimeBase component.

If you don't care about keeping the actual time offset in sync, only about the velocity, you can just drive  it and update locally too. The time offset will diverge between different users, but if that's ok for your use-case that'll be the most efficient way to do it.

**Comment by 3x1t-5tyl3** [comment]
> 
> 
> There's a few ways this could be approached. How often are you varying the speed? If you only change it occasionally (e.g. on Impulse) a modified version of Stopwatch node for example with Speed input would work well, with minimal added network traffic.
> 
> Alternatively we can add node that does sync the accumulated value, but this might be a bit trickier, as it'll depend on the host handling that (or another user) and will depend on their update rate without any sort of smoothing or adjusting, though that's relatively easy to fix. There are mechanism in Neos that do this kind of thing, like AuthorityTimeBase component.
> 
> If you don't care about keeping the actual time offset in sync, only about the velocity, you can just drive it and update locally too. The time offset will diverge between different users, but if that's ok for your use-case that'll be the most efficient way to do it.

Synced definitely. Having it be unsynced -while more performant- can have adverse effects which are undesired.

Example: If the multiplier is set to 1.0 after 10 seconds the value is 10. If I change the multiplier to 0.1 after ten seconds the value is 11. I make the multiplier 5.0 and after 10 seconds the value would be 61. (10 from the first time, 1 from the second and 50 from the third change.)

**Comment by Frooxius** [comment]
I understand your request, what I'm offering are different ways to tackle this, which each have somewhat different quirks and network loads and suit different patterns of usage. Knowing those can help, e.g. do you vary the speed value every frame or does it only occasionally change?

**Comment by 3x1t-5tyl3** [comment]
For most of my uses I change it every frame.

**Comment by StarfishHidari** [comment]
this is something that would help me a lot as incremental modules like the one posted by shadowpanther are used constantly in the things I'm making, I am typically varying the speed every frame as I'm driving the speed off of volume

**Comment by Earthmark** [comment]
If I understand correctly, this use case is for per frame differences in rate, but the value to stay synchronized?

I am curious about the edge cases here, based on your usages what do you feel should happen in these situations:

If  the rate is different between two users, one's rate is -1 and the other's is 1, what value should the node return? What should it return after a few seconds?

If there is a difference in output value that needs to be synchronized, say user A is at 1, user B is at 2, and say they somehow come to a sync agreement that they need to both be at 1 (so A is the chosen one, it's probably host), what should B do?
